### PR TITLE
SONAR-8333 4 log files

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/app/ServerProcessLogging.java [hunk 1]
```diff
@@ -52,12 +59,24 @@ public LoggerContext configure(Props props) {
...
+    PatternLayoutEncoder fileEncoder = new PatternLayoutEncoder();
+    fileEncoder.setContext(ctx);
+    fileEncoder.setPattern(logFormat);
+    fileEncoder.start();
+    fileAppender.setEncoder(fileEncoder);
+    fileAppender.start();
 
-  protected abstract void configureAppenders(String logFormat, LoggerContext ctx, LogbackHelper helper, Props props);
+    // configure logger
+    Logger rootLogger = ctx.getLogger(ch.qos.logback.classic.Logger.ROOT_LOGGER_NAME);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/app/ServerProcessLogging.java]: MAJOR Make the "rootLogger" logger private static final and rename it to comply with the format "LOG(?:GER)?". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1312)
MINOR Replace this fully qualified name with "Logger" [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1942)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: server/sonar-process/src/main/java/org/sonar/process/LogbackHelper.java [hunk 2]
```diff
@@ -73,6 +76,25 @@ public LoggerContextListener enableJulChangePropagation(LoggerContext loggerCont
...
+    fileAppender.setContext(ctx);
+    PatternLayoutEncoder fileEncoder = new PatternLayoutEncoder();
+    fileEncoder.setContext(ctx);
+    fileEncoder.setPattern(logFormat);
+    fileEncoder.start();
+    fileAppender.setEncoder(fileEncoder);
+    fileAppender.start();
+
+    // configure logger
+    Logger rootLogger = ctx.getLogger(ch.qos.logback.classic.Logger.ROOT_LOGGER_NAME);
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/LogbackHelper.java]: MAJOR Make the "rootLogger" logger private static final and rename it to comply with the format "LOG(?:GER)?". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1312)
MINOR Replace this fully qualified name with "Logger" [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1942)

File: server/sonar-process/src/main/java/org/sonar/process/LogbackHelper.java [hunk 3]
```diff
@@ -73,23 +87,207 @@ public LoggerContextListener enableJulChangePropagation(LoggerContext loggerCont
...
+   *     the max number of files defined in property {@link #MAX_FILES_PROPERTY}</li>
+   *     <li>the logs will follow the specified log pattern</li>
+   *   </ul>
+   * </p>
+   *
+   * @see #buildLogPattern(RootLoggerConfig)
+   */
+  public FileAppender<ILoggingEvent> configureGlobalFileLog(Props props, RootLoggerConfig config, String logPattern) {
+    LoggerContext ctx = getRootContext();
+    Logger rootLogger = ctx.getLogger(ROOT_LOGGER_NAME);
```
Comment (SonarTech) [inline server/sonar-process/src/main/java/org/sonar/process/LogbackHelper.java]: MAJOR Make the "rootLogger" logger private static final and rename it to comply with the format "LOG(?:GER)?". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1312)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by simonbrandhof** [comment]
As expected `sonar.log` contains only the bootstrapper logs about processes:
```
--> Wrapper Started as Daemon
Launching a JVM...
Wrapper (Version 3.2.3) http://wrapper.tanukisoftware.org
  Copyright 1999-2006 Tanuki Software, Inc.  All Rights Reserved.

2016.11.15 14:31:20 INFO  app[][o.s.a.AppFileSystem] Cleaning or creating temp directory /Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/temp
2016.11.15 14:31:20 INFO  app[][o.s.p.m.JavaProcessLauncher] Launch process[es]: /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/bin/java -Djava.awt.headless=true -Xmx1G -Xms256m -Xss256k -Djna.nosys=true -XX:+UseParNewGC -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupancyFraction=75 -XX:+UseCMSInitiatingOccupancyOnly -XX:+HeapDumpOnOutOfMemoryError -Djava.io.tmpdir=/Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/temp -javaagent:/Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/lib/management-agent.jar -cp ./lib/common/*:./lib/search/* org.sonar.search.SearchServer /Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/temp/sq-process7371156711751705287properties
objc[10623]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/bin/java and /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/lib/libinstrument.dylib. One of the two will be used. Which one is undefined.
2016.11.15 14:31:26 INFO  app[][o.s.p.m.Monitor] Process[es] is up
2016.11.15 14:31:26 INFO  app[][o.s.p.m.JavaProcessLauncher] Launch process[web]: /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/bin/java -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Djruby.management.enabled=false -Djruby.compile.invokedynamic=false -Xmx512m -Xms128m -XX:+HeapDumpOnOutOfMemoryError -Djava.io.tmpdir=/Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/temp -javaagent:/Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/lib/management-agent.jar -cp ./lib/common/*:./lib/server/*:/Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/lib/jdbc/h2/h2-1.3.176.jar org.sonar.server.app.WebServer /Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/temp/sq-process177472877338072998properties
objc[10626]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/bin/java and /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/lib/libinstrument.dylib. One of the two will be used. Which one is undefined.
2016.11.15 14:31:43 INFO  app[][o.s.p.m.Monitor] Process[web] is up
2016.11.15 14:31:43 INFO  app[][o.s.p.m.JavaProcessLauncher] Launch process[ce]: /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/bin/java -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Xmx512m -Xms128m -XX:+HeapDumpOnOutOfMemoryError -Djava.io.tmpdir=/Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/temp -javaagent:/Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/lib/management-agent.jar -cp ./lib/common/*:./lib/server/*:./lib/ce/*:/Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/lib/jdbc/h2/h2-1.3.176.jar org.sonar.ce.app.CeServer /Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/temp/sq-process257269035489269671properties
objc[10629]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/bin/java and /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/lib/libinstrument.dylib. One of the two will be used. Which one is undefined.
2016.11.15 14:31:47 INFO  app[][o.s.p.m.Monitor] Process[ce] is up
TERM trapped.  Shutting down.
2016.11.15 14:32:28 INFO  app[][o.s.p.m.Monitor] Process[ce] is stopping
2016.11.15 14:32:30 INFO  app[][o.s.p.m.Monitor] Process[ce] is stopped
2016.11.15 14:32:30 INFO  app[][o.s.p.m.Monitor] Process[web] is stopping
2016.11.15 14:32:33 INFO  app[][o.s.p.m.Monitor] Process[web] is stopped
2016.11.15 14:32:33 INFO  app[][o.s.p.m.Monitor] Process[es] is stopping
2016.11.15 14:32:34 INFO  app[][o.s.p.m.Monitor] Process[es] is stopped
<-- Wrapper Stopped
```

It's not trivial to know when server is effectively started and stopped. As bootstrapper knows which processes are enabled, could it log the synthesis "SonarQube is started/stopped" ?

**Comment by sns-seb** [comment]
> It's not trivial to know when server is effectively started and stopped. As bootstrapper knows which processes are enabled, could it log the synthesis "SonarQube is started/stopped" ?

Hello @simonbrandhof , please meet https://jira.sonarsource.com/browse/SONAR-8341, I think you and it are meant for each other :)

**Comment by simonbrandhof** [comment]
When a process crashes at startup, there are no ERROR logs. The error message is displayed in "System.out" and has no time nor other fields:
```
2016.11.15 14:49:05 INFO  app[][o.s.p.m.JavaProcessLauncher] Launch process[ce]: /Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/bin/java -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Xmx1m -XXX:+HeapDumpOnOutOfMemoryError -Djava.io.tmpdir=/Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/temp -javaagent:/Library/Java/JavaVirtualMachines/jdk1.8.0_91.jdk/Contents/Home/jre/lib/management-agent.jar -cp ./lib/common/*:./lib/server/*:./lib/ce/*:/Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/lib/jdbc/h2/h2-1.3.176.jar org.sonar.ce.app.CeServer /Users/sbrandhof/dev/core/sonarqube/sonar-application/target/sonarqube-6.2-SNAPSHOT/temp/sq-process8835893891066495281properties
Unrecognized option: -XXX:+HeapDumpOnOutOfMemoryError
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
2016.11.15 14:49:05 INFO  app[][o.s.p.m.Monitor] Process[web] is stopping
2016.11.15 14:49:08 INFO  app[][o.s.p.m.Monitor] Process[web] is stopped
```

**Comment by simonbrandhof** [comment]
FYI tested successfully file rotation + change of log level + change of path

File: server/sonar-server/src/main/java/org/sonar/ce/log/CeActivityLogAcceptFilter.java [hunk 4]
```diff
@@ -1,40 +0,0 @@
...
-package org.sonar.ce.log;
-
-import ch.qos.logback.core.filter.Filter;
-import ch.qos.logback.core.spi.FilterReply;
-import org.slf4j.MDC;
-
-/**
- * Keeps only Compute Engine activity logs.
- */
-public class CeActivityLogAcceptFilter<E> extends Filter<E> {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/ce/log/CeActivityLogAcceptFilter.java]: Good news that CE logback filters are dropped.

File: sonar-application/src/main/java/org/sonar/application/AppLogging.java [hunk 5]
```diff
@@ -19,85 +19,201 @@
...
-    fileEncoder.setContext(ctx);
-    fileEncoder.setPattern("%msg%n");
-    fileEncoder.start();
-    fileAppender.setEncoder(fileEncoder);
-    fileAppender.start();
-
-    // configure logger
-    Logger gobblerLogger = ctx.getLogger(GOBBLER_LOGGER);
+  /**
+   * The process has been started by orchestrator (ie. via {@code java -jar} and optically passing the option {@code -Dsonar.log.console=true}).
```
Comment (simonbrandhof) [inline sonar-application/src/main/java/org/sonar/application/AppLogging.java]: "optically" ?

**Comment by sns-seb** [comment]
> When a process crashes at startup, there are no ERROR logs. The error message is displayed in "System.out" and has no time nor other fields:

I've added a INFO log message with the key of the process indicating that it failed to start

**Comment by sns-seb** [comment]
> when using property "sonar.path.logs=logs2", the file sonar.log is lost. The directory logs2 contains only ce.log, es.log and web.log.

This can't be solved in App as the writing to sonar.log is actually handled by the wrapper

**Pull request merged**